### ISSUE TITLE
Load Libraries Using Dynamic TLS

### ISSUE THREAD

**Issue by lilleswing** [mdtraj/mdtraj#1324]
Can we build a version of mdtraj which dynamically loads the dll's instead of the fast static loading?

[DeepChem](https://github.com/deepchem/deepchem) uses mdtraj for loading in and cleaning up data.

After upgrading tensorflow to 1.4 we run have out of things that we can open with static TLS, on travis.
https://travis-ci.org/deepchem/deepchem/jobs/312146212.  
Travis uses Ubuntu 14.04 which is linux kernel 3.something, the linux kernel raised the limits sometime after this kernel version.  Travis has no plans on upgrading the OS of their CI boxes before 14.04 EOL.

@peastman suggested adding the compile flags -fPIC which I hard-coded here [github](https://github.com/lilleswing/mdtraj/pull/1/files) [anaconda cloud](https://anaconda.org/deepchem/mdtraj).  However that did not stop us from still going over the limit.

tl;dr
Are there compile time flags I can use for mdtraj to use dynamic TLS instead of static TLS?

**Comment by rmcgibbo** [comment]
I think I know what you're talking about, but I'm not familiar with some of the terminolgy that you're using ("dynamic TLS"). [Googling that term](https://www.google.com/search?q="dynamic+tls"&oq="dynamic+tls") yields hits that are all referring to transport level security.

**Comment by rmcgibbo** [comment]
But to answer the question I think you're asking, it adds a _ton_ of complexity to dlopen all of the run time dynamic libraries instead of compiling in the standard way, with the libraries referenced (on linux) in the ELF dynamic section with `DT_NEEDED`. That's not something that's in the cards for MDTraj.

**Comment by lilleswing** [comment]
I actually don't know a lot about it.  The most informative page I could find was this
https://bugzilla.redhat.com/show_bug.cgi?id=1124987
Which is talking about fedora in 2014.
And this google groups page
https://groups.google.com/a/tensorflow.org/forum/#!topic/discuss/KPUUdK2vA9Q

Here is an example error I am seeing
```bash
Traceback (most recent call last):
  File "/home/travis/build/deepchem/deepchem/deepchem/feat/tests/test_atomic_coordinates.py", line 77, in test_neighbor_list_extremes
    nblist = nblist_featurizer._featurize(self.mol)[1]
  File "/home/travis/build/deepchem/deepchem/deepchem/feat/atomic_coordinates.py", line 147, in _featurize
    self.periodic_box_size)
  File "/home/travis/build/deepchem/deepchem/deepchem/feat/atomic_coordinates.py", line 58, in compute_neighbor_list
    import mdtraj
  File "/home/travis/anaconda/envs/deepchem/lib/python2.7/site-packages/mdtraj/__init__.py", line 53, in <module>
    from mdtraj._rmsd import rmsd
ImportError: dlopen: cannot load any more object with static TLS
```

**Comment by lilleswing** [comment]
If there is nothing you can do thanks for looking at it.

**Comment by rmcgibbo** [comment]
Okay, this is somehow related to limits on the amount of thread-local storage.

I don't understand why you suspect this would be remedied by "a version of mdtraj which dynamically loads the dll's instead of the fast static loading." Could you explain what you have in mind more?

**Comment by rmcgibbo** [comment]
Also: a bit of googling of the query "cannot load any more object with static TLS" suggest pretty strongly to me that the bad actor here, if any, is tensorflow or pytorch. To the extent that there are process-wide global limits on the amount of thread local storage, it's the deep learning libraries that are gobbling up most of the limit. I strongly suspect that MDTraj's rmsd module uses only a little bit.

**Comment by rmcgibbo** [comment]
You could probably reduce the amount of thread-local storage used my mdtraj to zero by compiling the package with `python setup.py install --disable-openmp`, but that's not how the anaconda binaries will be compiled.

**Comment by lilleswing** [comment]
I agree it is the Deep learning library doing something bad (lots of messages about pytorch and tensorflow with this issue), but I don't think I can do too much about those :(

I'm trying with --disable-openmp now

**Comment by rmcgibbo** [comment]
If you compile tensortorch yourself from scatch, you might get something (vs using precompied binaries)

**Comment by rmcgibbo** [comment]
Closing, since I don't think there's anything we can do at the MDTraj project level.
